Fix JBMC crash on windows in `NondetEnumOpaqueReturn` test

thomasspriggs opened the pull request:
String literals with multibyte utf-8 characters in java cause `ch` to be
negative. Passing a negative value into `std::isalnum` results in
undefined behaviour. On windows the observed behaviour is a popup error
message about a failed assertion in the run time library. This can stall
the CI process as none of the buttons on the popup message ever get
pressed. Escaping these characters is sufficient to fix the stalling
test.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] N/A - Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] N/A - The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] N/A - My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

martin-cs commented on src/util/string_utils.cpp:
```diff
@@ -158,7 +158,7 @@ std::string escape_non_alnum(const std::string &to_escape)
     if(ch == '_')
       escaped << "__";
-    else if(isalnum(ch))
+    else if(ch >= 0 && isalnum(ch))
```
I'm a little worried about this because certainly in C whether `char` is signed or not is implementation dependent.

thomasspriggs replied:
Now done through a `narrow_cast` cast, in order to avoid explicitly checking for negative values.

thomasspriggs committed:
Fix JBMC crash on windows in `NondetEnumOpaqueReturn` test

String literals with multibyte utf-8 characters in java cause `ch` to be
negative. Passing a negative value into `std::isalnum` results in
undefined behaviour. On windows the observed behaviour is a popup error
message about a failed assertion in the run time library. This stalls
the CI process as none of the buttons on the popup message ever get
pressed. Escaping these characters is sufficient to fix the stalling
test.

martin-cs reviewed (approved):
If the test passes on all platforms then I'm happy.

thomasspriggs committed:
Fix conversion to hex of negative characters

Negative characters were previously converted with unneccessary leading
`f` characters. For example `-1` or `0xff` characters were previously
printed as `_ffffffff`. This commit uses the value of characters in the
-128 to -1 range remapped into the 128 to 255 range, so that `-1` prints
as `_ff`, without unneccessary padding with leading `f`s.

thomasspriggs committed:
Add unit test of `escape_non_alnum`

To show that this function now works correctly with all possible one
byte strings.

martin-cs reviewed (approved):
This seems like a better way of doing things.  If possible it might be good to squash some of the commits together.

hannes-steffenhagen-diffblue commented on src/util/string_utils.cpp:
```diff
@@ -156,13 +156,16 @@ std::string escape_non_alnum(const std::string &to_escape)
   std::ostringstream escaped;
   for(auto &ch : to_escape)
   {
+    const auto uch = static_cast<unsigned char>(ch);
```
Actually had to look this up, the standard actually guarantees this is well defined and behaves like a sane person would expect! A surprise to be sure, but a welcome one.

hannes-steffenhagen-diffblue commented on unit/util/string_utils/escape_non_alnum.cpp:
```diff
@@ -0,0 +1,278 @@
+  CHECK(escape_non_alnum(std::string{static_cast<signed char>(0xFC)}) == "_fc");
+  CHECK(escape_non_alnum(std::string{static_cast<signed char>(0xFD)}) == "_fd");
+  CHECK(escape_non_alnum(std::string{static_cast<signed char>(0xFE)}) == "_fe");
+  CHECK(escape_non_alnum(std::string{static_cast<signed char>(0xFF)}) == "_ff");
```
I hope this is generated instead of handwritten!

hannes-steffenhagen-diffblue commented on src/util/string_utils.cpp:
```diff
@@ -156,13 +156,16 @@ std::string escape_non_alnum(const std::string &to_escape)
+    {
       escaped << '_' << std::hex << std::setfill('0') << std::setw(2)
-              << (unsigned int)ch;
+              << (unsigned int)uch;
```
Since you’re changing this line anyway could you turn this into `static_cast` please?

thomasspriggs replied:
Now done as a `narrow` style cast instead as that is the preferred way to cast numerics according to the C++ core guidelines.

hannes-steffenhagen-diffblue replied:
@thomasspriggs I don't think this is a good in this instance as this is a widening cast, not a narrowing cast, so it's a bit confusing. Not a blocker, just an odd thing.

thomasspriggs replied:
Ok. I have swapped the conversion to `unsigned char` back to `static_cast`, because reinterpretation is the desired behaviour here. I have swapped the conversion from `unsigned char` to `int`, such that it is done using `{}` braced initialisation because this form of initialisation supports widening, but not narrowing. Therefore it makes it explicit that we are not performing narrowing.

thomasspriggs commented:
> This seems like a better way of doing things. If possible it might be good to squash some of the commits together.

Now squashed. This was still in draft status as I was still tidying things up.

thomasspriggs committed:
Remove usage of `narrow` for widening conversion

hannes-steffenhagen-diffblue commented on src/util/string_utils.cpp:
```diff
@@ -156,13 +156,13 @@ std::string escape_non_alnum(const std::string &to_escape)
   std::ostringstream escaped;
   for(auto &ch : to_escape)
   {
+    const int uch{static_cast<unsigned char>(ch)};
```
Nit: This is now complicated enough to warrant an explanation I'd think. Would be great if you could drop a line here of why this is done.

thomasspriggs replied:
Done.

thomasspriggs committed:
Explain why particular conversions of characters are performed

Because the reasons are subtle and non-obvious.

NlightNFotis reviewed (approved):
LGTM
